feat(aio): add attribute utils for code atty interpretation.

wardbell opened the pull request:
These utils support flexible, natural attribute interpretation as applied to code-example and code-pane. Then apply those utils to `code-example` and `live-example`.

Second commit leverages the attribute utils changes to style "avoid" code-examples by convention.
Keep with this PR to avoid merge conflict

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

wardbell committed:
feat(aio): add attribute utils for code atty interpretation.

These utils support flexible, natural attribute interpretation as applied to code-example and code-pane. Then apply those utils to code-example and live-example

mary-poppins commented:
The angular.io preview for 63e0183c811e3934d82b0fc929fb5e8cc59c7320 is available [here][1].

[1]: https://pr16496-63e0183c811e3934d82b0fc929fb5e8cc59c7320.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 14646ad475055be2a78b74956482dc153ee2ef4f is available [here][1].

[1]: https://pr16496-14646ad475055be2a78b74956482dc153ee2ef4f.ngbuilds.io/

wardbell committed:
feat(aio): set “avoid” class by convention (filename has “.avoid.”)

Most important for StyleGuide which lost example css class in migration.
Also hides copy for “avoid” files.

mary-poppins commented:
The angular.io preview for f0cd1833cf7ba56b7cfde8b4681a637e6cd4ca66 is available [here][1].

[1]: https://pr16496-f0cd1833cf7ba56b7cfde8b4681a637e6cd4ca66.ngbuilds.io/

petebacondarwin commented on aio/src/app/shared/attribute-utils.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+/**
+ * Return the attribute that matches `atty`.
```
atty -> attr

petebacondarwin commented on aio/src/app/shared/attribute-utils.ts:
```diff
@@ -0,0 +1,48 @@
+
+/**
+ * Return the boolean state of an attribute value (if supplied)
+ * @param attyValue The string value of some attribute (or undefined if attribute not present)
```
attyValue -> attrValue

petebacondarwin commented on aio/src/app/shared/attribute-utils.spec.ts:
```diff
@@ -0,0 +1,148 @@
+    testEl = div.querySelector('div');
+  });
+
+  describe('#getAttrs', () => {
```
pretty sure that the `#` character is only used to denote a member property or method. These are all free standing export functions.

wardbell replied:
Good point.

petebacondarwin commented on aio/src/app/shared/attribute-utils.spec.ts:
```diff
@@ -0,0 +1,148 @@
+    });
+
+    it('should return empty string value for attribute "A"', () => {
+      expect(getAttrValue(attrMap, 'a')).toBe('');
```
`a` -> `A`? Otherwise this is the same as the previous test.

petebacondarwin commented on aio/src/app/shared/attribute-utils.spec.ts:
```diff
@@ -0,0 +1,148 @@
+    });
+
+    it('should return empty string value for attribute "d-E"', () => {
+      expect(getAttrValue(attrMap, 'd-e')).toBe('');
```
`d-e` -> `d-E`? Or is the test description incorrect?

petebacondarwin commented on aio/src/app/shared/attribute-utils.spec.ts:
```diff
@@ -0,0 +1,148 @@
+    });
+
+    it('should return empty string for attribute ["d-e"]', () => {
+      // because d-e will be found before d
```
I don't think this comment meant to be here.

petebacondarwin commented on aio/src/app/shared/attribute-utils.spec.ts:
```diff
@@ -0,0 +1,148 @@
+    });
+
+    it('should return true for present but unassigned attr "a"', () => {
+      expect(boolFromValue(getAttrValue(attrMap, 'a'))).toBe(true);
```
What is the point of using `getAttrValue()` throughout the tests in this describe block? It just makes it more difficult to see what the behaviour of the function is. Why not just use the value directly? E.g.

```
expect(boolFromValue('')).toBe(true);
```

petebacondarwin replied:
The combination of the two functions are effectively tested below in the `getBoolFromAttribute()` tests.

petebacondarwin commented on aio/src/app/shared/attribute-utils.ts:
```diff
@@ -0,0 +1,48 @@
+ * @param atty Name of the attribute or a string of candidate attribute names
+ */
+export function getAttrValue(attrs: StringMap, attr: string | string[] = ''): string {
+  return attrs[typeof attr === 'string' ? attr : attr.find(a => attrs[a] !== undefined)];
```
I don't believe that `array.find` works on IE - do we have a polyfill in place for that?

wardbell replied:
We better. All kinds of things depend on that.

gkalpak replied:
No, the app doesn't work on IE currently. We don't include the necessary polyfills.

petebacondarwin reviewed (approved):
LGTM except for the minor test issues, which I am happy to fix up in a follow up PR.

mary-poppins commented:
The angular.io preview for 81e186df4b244f3ddfbb133fcd79893ed07d58fd is available [here][1].

[1]: https://pr16496-81e186df4b244f3ddfbb133fcd79893ed07d58fd.ngbuilds.io/
